### PR TITLE
ObjectMeta, ListMeta, and TypeMeta should implement their interfaces

### PR DESCRIPTION
**Pull request by smarterclayton**
Make unversioned.ListMeta implement List. Update all the *List types so they implement GetListMeta.
This helps avoid using reflection to get list information.

Remove all unnecessary boilerplate, move the interfaces to the right
places, and add a test that verifies that objects implement one, the
other, but never both.

@ncdc @lavalamp this supercedes #26964 with the boilerplate removed.  Added tests

### CODE REVIEW

File: pkg/apis/apps/register.go [hunk 1]
```diff
@@ -53,5 +54,7 @@ func addKnownTypes(scheme *runtime.Scheme) {
 	)
 }
 
+func (obj *PetSet) GetObjectMeta() meta.Object                { return &obj.ObjectMeta }
 func (obj *PetSet) GetObjectKind() unversioned.ObjectKind     { return &obj.TypeMeta }
 func (obj *PetSetList) GetObjectKind() unversioned.ObjectKind { return &obj.TypeMeta }
+func (obj *PetSetList) GetListMeta() meta.List                { return &obj.ListMeta }
```
Comment (lavalamp) [inline pkg/apis/apps/register.go]: Missed this file?

**Comment by lavalamp** [comment]
Awesome! Just one comment-- self LGTM when you fix it.

**Comment by smarterclayton** [comment]
Fixed

**Comment by smarterclayton** [comment]
@k8s-bot test this issue #25635

**Comment by smarterclayton** [comment]
```
build/make-clean.sh
+++ [0610 10:34:55] Verifying Prerequisites....
+++ [0610 10:34:55] Cleaning out _output/dockerized/bin/ via docker build image
+++ [0610 10:34:55] Running build command....
+++ [0610 10:34:59] Removing data container
+++ [0610 10:35:05] Cleaning out local _output directory
+++ [0610 10:35:05] Deleting docker image kube-build:build-f9b01f6cbf
Untagged: kube-build:build-f9b01f6cbf
Deleted: db22dc42cd56a16005fac3c7377ad4afc0c650e8b1e804db657b5d0fa90c6693
Deleted: ed6727b8ea034b7667fe75384c2b61a73531452cb03cbefed7b690c2ec7aa51f
Deleted: bf89151b8f9ca3592146a90535a7fa00edad5d4b6a45833316301a78a6cf05fe
Deleted: e6b3af5f4168807cc45f05f65922a5b8581e6b44f0f640c77acc86b39b4a2742
+++ [0610 10:35:08] Cleaning all other untagged docker images
rm -rf _output
rm -rf Godeps/_workspace # Just until we are sure it is gone
+ git clean -fdx
+ go run ./hack/e2e.go -v --build
2016/06/10 10:35:09 e2e.go:212: Running: build-release
ERROR: gcloud crashed (error): [Errno 111] Connection refused

If you would like to report this issue, please run the following command:
  gcloud feedback
2016/06/10 10:35:10 e2e.go:218: Error running build-release: exit status 1
2016/06/10 10:35:10 e2e.go:214: Step 'build-release' finished in 1.14318742s
2016/06/10 10:35:10 e2e.go:95: Error building. Aborting.
```

Connection refused is probably a true flake (related to GCP flakiness since last night?).  Will spawn a new issue if it recurs.

**Comment by smarterclayton** [comment]
@k8s-bot test this issue #26799

**Comment by smarterclayton** [comment]
Broke due to #27205

**Comment by smarterclayton** [comment]
@k8s-bot test this issue #27205

**Pull request merged**